index.d.ts

```diff
@@ -127,21 +131,54 @@ export namespace Cmd {
           ? typeof dispatch
           : T[K];
   }
+
   export type PromiseResult<T> = T extends Promise<infer U> ? U : T;
 
+  type RunFunc = (...args: any[]) => Promise<any> | any;
+
+  type RunOptions<
+    Func extends RunFunc,
+    SuccessAction extends Action = never,
+    FailAction extends Action = never,
+    FailReason = unknown
+  > = {
+    args?: ArgOrSymbol<Parameters<Func>>;
+    forceSync?: boolean;
+    testInvariants?: boolean;
+    successActionCreator: (value: PromiseResult<ReturnType<Func>>) => SuccessAction;
+    failActionCreator: (error: FailReason) => FailAction;
+  }
+
+  export function run<Func extends RunFunc>(
+    f: Func,
+    options?: Omit<RunOptions<Func>, 'successActionCreator' | 'failActionCreator'>,
+  ): RunCmd;
+
   export function run<
     Func extends (...args: any[]) => Promise<any> | any,
     SuccessAction extends Action,
+  >(
+    f: Func,
+    options: Omit<RunOptions<Func, SuccessAction>, 'failActionCreator'>,
+  ): RunCmd<SuccessAction, never>;
+
+  export function run<
+    Func extends (...args: any[]) => Promise<any> | any,
     FailAction extends Action,
-    >(
+    FailReason = unknown
+  >(
     f: Func,
-    options?: {
-      args?: ArgOrSymbol<Parameters<Func>>;
-      failActionCreator?: (error: any) => FailAction;
-      successActionCreator?: (value: PromiseResult<ReturnType<Func>>) => SuccessAction;
-      forceSync?: boolean;
-      testInvariants?: boolean;
-    },
+    options: Omit<RunOptions<Func, never, FailAction, FailReason>, 'successActionCreator'>,
+  ): RunCmd<never, FailAction>;
+
+  export function run<
+    Func extends (...args: any[]) => Promise<any> | any,
+    SuccessAction extends Action = never,
```

Comment:
SuccessAction and FailAction should not have defaults here. If you don't specify either it can be one of the other overloads